gbfs-validator/gbfs.js

```diff
@@ -5,45 +5,155 @@ const validatorVersion = process.env.COMMIT_REF
   : require('./package.json').version
 
 function hasErrors(data, required) {
-  let hasError = false
-
-  data.forEach((el) => {
+  for (const el of data) {
     if (Array.isArray(el)) {
       if (hasErrors(el, required)) {
-        hasError = true
+        return true
       }
     } else {
-      if (required && !el.exists ? true : !!el.errors || el.hasErrors) {
-        hasError = true
+      if (required && !el.exists) {
+        return true
+      }
+
+      if (el.hasErrors || el.errors?.length || el.nonSchemaErrors?.length) {
+        return true
       }
     }
-  })
+  }
 
-  return hasError
+  return false
+}
+
+function hasWarnings(data) {
+  for (const el of data) {
+    if (Array.isArray(el)) {
+      if (hasWarnings(el)) {
+        return true
+      }
+    }
+
+    if (el.hasWarnings || el.nonSchemaWarnings?.length) {
```

Comment:
Since I was not able to easily combine `errors` and `nonSchemaErrors`, I thought it was best to keep the same logic for the `warnings`. Anyone, feel free to suggest an implementation.